features/332-sso/api.md

```diff
@@ -0,0 +1,311 @@
+```
+
+### Example
+
+```typescript
+askConfirmation().then(confirmed => {
+  if (confirmed) {
+    skygear.auth.removeOAuthProvider("google");
+  }
+});
+```
+
+# Merging existing user for login
+
+Some users may forget that they have signed up with email long ago. When they return to the application again, they may login with OAuth 2.0, creating a new user identity unexpectedly. To mitigate this issue, the developer can instruct the Auth gear on how to merge existing user.
+
+All login APIs accept `LoginOptions`. By default, merging is enabled. If `mergeRealm` is not in `allowedRealms`, it is an error and the login process is aborted. Otherwise, the implementation will merge existing user by inferring email from the provider user info and use it and the realm to match existing user.
+
+> In addition to email, should we consider phone number?
+
+# Removal of v1 API
+
+The following v1 API is removed.
+
+```typescript
+function getOAuthProviderProfiles(): Promise<{[providerID: string]: any}>;
+```
+
+The existing implementation simply return all snapshots of user info of all providers of the logged in user. The snapshot of user info most likely will be stale. So the function unlikely return what the developer expects. If we want to make the function always return the fresh user info, we need to implement refresh token of OAuth 2.0. However, to implement refresh token, we have to drop support for `withAccessToken`, not to mention that some providers like Facebook does not support refresh token.
+
+# List of Supported OAuth 2.0 Providers
+
+- Google
+- Facebook
+- Instagram
+- LinkedIn
+- [Microsoft identify platform](https://docs.microsoft.com/en-us/azure/active-directory/develop/v2-oauth2-auth-code-flow)
+
+# Custom Token
+
+In case the developer want to integrate with a provider who does not support OAuth 2.0, they can achieve single sign-on with Custom Token.
+
+### Client Side API
+
+```typescript
+function loginWithCustomToken(token: string, options?: LoginOptions): Promise<User>;
+```
+
+### Client Side Example
+
+```typescript
+loginMyAuthenticationServer().then(token => {
+  skygear.auth.loginWithCustomToken(token).then(user => {
+    // login suceeded
+  });
+});
+```
+
+### Server Side Integration
+
+The custom token **MUST** be a JWT token.
+
+It **MUST** have the following standard claims.
+
+- `iss`: The value must match the configured value in tenant configuration.
+- `sub`: The user ID of the foreign authentication system.
+
+It **SHOULD** have the following standard claims.
+
+- `aud`: If present and the configured value is not empty, they must match.
+- `exp`: If present, the Auth gear will validate it against the current time.
+- `nbf`: If present, the Auth gear will validate it against the current time.
+
+The signing algorithm is `HS256`. The signing key is shared between the Auth gear and the foreign authentication system.
+
+The token can optionally contain a claim named `skyprofile` which **MUST** be an JSON object.
```

Comment:
Can we use another name? For email, we can use OIDC standard JWT claim `email`. For custom profile, we can use prefixed name like `skygear:profile` instead.

Reply:
This is actually following the old spec. So you are proposing we lift `email` up a level so it become a top-level claim? Other than `email`, we are not doing anything special with `skyprofile` so probably we can drop it.